registration/static/js/src/registration.js

```diff
@@ -54,10 +53,45 @@ app.controller('Registration', ['$scope', '$http', 'djangoForm', function($scope
         return getFormFields().filter(key => $scope.form[key].$dirty || $scope.form[key].$message);
     };
 
+    // Returns the result of validating password using zxcvbn.
+    var validatePassword = function() {
+        var password = $scope.form.password.$viewValue;
+        return zxcvbn(password);
+    };
+
+    // Makes sure each suggestion ends with period.
+    var formatSuggestions = function(suggestions) {
+        return _.map(suggestions, function(suggestion) {
+            if (!suggestion.endsWith('.')) {
+                return suggestion + '.';
+            }
+            return suggestion;
+        });
+    };
+
+    // Returns feedback from password validation.
+    var getValidationFeedback = function() {
+        var validationResult = validatePassword(),
+            suggestions = validationResult.feedback.suggestions,
+            warning = validationResult.feedback.warning;
+        if (warning) {
+            suggestions.push(warning);
+        }
+        suggestions = formatSuggestions(suggestions);
+        // Concatenate suggestions to ensure that all of them are displayed
+        return [suggestions.join(" ")];
+    };
+
     // Display the given error messages, making sure not to clear messages
     // that are already displayed.
     var displayErrors = function(errors) {
         errors = _.pick(errors, getModifiedFields());
+        if (errors.password && $scope.form.password.$viewValue) {
```

Comment:
You can validate this on the client directly without a roundtrip to the server:

- Remove this
- Remove server-side password strength validation (so that client-side validation messages are not replaced by the error message from the server)
- Remove 'password' from `serverValidationFields`
- Add something like this:

    ```javascript
    $scope.$watch('registration.password', function() {
        displayErrors({ password: getValidationFeedback() });
    });
    ```